packages/protocol/contracts/common/FeeHandler.sol

```diff
@@ -355,20 +451,150 @@ contract FeeHandler is
+  // keep track of amounts to be burned, like the Celo token does
+  function _setDistributionAmounts(
+    TokenState storage tokenState,
+    IERC20 token
+  ) internal returns (uint256) {
+    uint256 balanceOfToken = token.balanceOf(address(this));
+    uint256 balanceToProcess = balanceOfToken.sub(tokenState.toDistribute);
 
-  function _setFeeBeneficiary(address beneficiary) private {
-    feeBeneficiary = beneficiary;
-    emit FeeBeneficiarySet(beneficiary);
+    uint256 balanceToBurn = _setDistributeAfterBurn(tokenState, balanceToProcess);
+
+    emit DistributionAmountSet(address(token), tokenState.toDistribute);
+    return balanceToBurn;
   }
 
-  function _setBurnFraction(uint256 newFraction) private {
-    FixidityLib.Fraction memory fraction = FixidityLib.wrap(newFraction);
+  function _executePayment(
+    address tokenAddress,
+    TokenState storage state,
+    address beneficiary,
+    uint256 amount
+  ) internal {
     require(
-      FixidityLib.lte(fraction, FixidityLib.fixed1()),
-      "Burn fraction must be less than or equal to 1"
+      _getTokenActive(tokenAddress) ||
+        tokenAddress == registry.getAddressForOrDie(CELO_TOKEN_REGISTRY_ID),
+      "Token needs to be active"
     );
-    burnFraction = fraction;
-    emit BurnFractionSet(newFraction);
+
+    IERC20 token = IERC20(tokenAddress);
+    uint256 balanceToDistribute = amount;
+
+    if (balanceToDistribute == 0) {
+      // don't distribute with zero balance
+      return;
+    }
+
+    token.transfer(beneficiary, balanceToDistribute);
+  }
+
+  function _setDistributeAfterBurn(
+    TokenState storage tokenState,
+    uint256 balanceToProcess
+  ) internal returns (uint256) {
+    uint256 balanceToBurn = FixidityLib
+      .newFixed(balanceToProcess)
+      .multiply(getBurnFractionFixidity())
+      .fromFixed(); //here2
+    tokenState.toDistribute = tokenState.toDistribute.add(balanceToProcess.sub(balanceToBurn));
+    return balanceToBurn;
+  }
+
+  function checkTotalBeneficiary() internal {
+    require(
+      getTotalFractionOfOtherBeneficiariesAndCarbonFixidity().lt(FixidityLib.fixed1()),
+      "Total beneficiaries fraction must be less than 1"
+    );
+  }
+
+  function _setCarbonFraction(uint256 _newFraction) internal {
+    FixidityLib.Fraction memory newFraction = FixidityLib.wrap(_newFraction);
+    require(newFraction.lt(FixidityLib.fixed1()), "New cargon fraction can't be greather than 1"); // TODO function
+    ignoreRenaming_inverseCarbonFraction = FixidityLib.fixed1().subtract(newFraction);
+    checkTotalBeneficiary();
+    emit CarbonFractionSet(_newFraction);
+  }
+
+  function _setBeneficiaryFraction(
+    address beneficiaryAddress,
+    FixidityLib.Fraction memory newFraction
+  ) internal {
+    Beneficiary storage beneficiary = otherBeneficiaries[beneficiaryAddress];
+    require(beneficiary.exists, "Beneficiary not found");
+    totalFractionOfOtherBeneficiaries = totalFractionOfOtherBeneficiaries.add(newFraction);
```

Comment:
We should be checking whether this is <= 1

Reply:
ok checkTotalBeneficiary is doing that - should we have it <= 1 rather than <1 ?

Reply:
I don't have tested the edge case where burningFraction = 0, so would leave it like this.